# --- T2-COPYRIGHT-BEGIN ---
# t2/package/*/firefox/hotfix-wgpu-atomicu64.patch
# Copyright (C) 2024 - 2026 The T2 SDE Project
# SPDX-License-Identifier: GPL-2.0 or patched project license
# --- T2-COPYRIGHT-END ---

Fix wgpu to build for 32-bit arch w/o 64-bit atomics, too.

--- firefox-158.0/third_party/rust/ash/src/device.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/ash/src/device.rs	2026-10-13 19:47:59.034318390 +0200
@@ -676,7 +676,7 @@ impl Device {
 
     /// <https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetSemaphoreCounterValue.html>
     #[inline]
-    pub unsafe fn get_semaphore_counter_value(&self, semaphore: vk::Semaphore) -> VkResult<u64> {
+    pub unsafe fn get_semaphore_counter_value(&self, semaphore: vk::Semaphore) -> VkResult<usize> {
         let mut value = mem::MaybeUninit::uninit();
         (self.device_fn_1_2.get_semaphore_counter_value)(
             self.handle(),
--- firefox-158.0/third_party/rust/ash/src/extensions/khr/timeline_semaphore.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/ash/src/extensions/khr/timeline_semaphore.rs	2026-10-13 19:47:59.035108807 +0200
@@ -7,7 +7,7 @@ use core::mem;
 impl crate::khr::timeline_semaphore::Device {
     /// <https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetSemaphoreCounterValue.html>
     #[inline]
-    pub unsafe fn get_semaphore_counter_value(&self, semaphore: vk::Semaphore) -> VkResult<u64> {
+    pub unsafe fn get_semaphore_counter_value(&self, semaphore: vk::Semaphore) -> VkResult<usize> {
         let mut value = mem::MaybeUninit::uninit();
         (self.fp.get_semaphore_counter_value_khr)(self.handle, semaphore, value.as_mut_ptr())
             .assume_init_on_success(value)
--- firefox-158.0/third_party/rust/ash/src/extensions_generated.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/ash/src/extensions_generated.rs	2026-10-13 19:47:59.063214008 +0200
@@ -14711,7 +14711,7 @@ pub mod khr {
                         unsafe extern "system" fn get_semaphore_counter_value_khr(
                             _device: crate::vk::Device,
                             _semaphore: Semaphore,
-                            _p_value: *mut u64,
+                            _p_value: *mut usize,
                         ) -> Result {
                             panic!(concat!(
                                 "Unable to load ",
--- firefox-158.0/third_party/rust/ash/src/tables.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/ash/src/tables.rs	2026-10-13 19:47:59.064096778 +0200
@@ -3463,7 +3463,7 @@ impl DeviceFnV1_2 {
                 unsafe extern "system" fn get_semaphore_counter_value(
                     _device: crate::vk::Device,
                     _semaphore: Semaphore,
-                    _p_value: *mut u64,
+                    _p_value: *mut usize,
                 ) -> Result {
                     panic!(concat!(
                         "Unable to load ",
--- firefox-158.0/third_party/rust/ash/src/vk/definitions.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/ash/src/vk/definitions.rs	2026-10-13 19:47:59.074571869 +0200
@@ -12913,9 +12913,9 @@ pub struct D3D12FenceSubmitInfoKHR<'a> {
     pub s_type: StructureType,
     pub p_next: *const c_void,
     pub wait_semaphore_values_count: u32,
-    pub p_wait_semaphore_values: *const u64,
+    pub p_wait_semaphore_values: *const usize,
     pub signal_semaphore_values_count: u32,
-    pub p_signal_semaphore_values: *const u64,
+    pub p_signal_semaphore_values: *const usize,
     pub _marker: PhantomData<&'a ()>,
 }
 unsafe impl Send for D3D12FenceSubmitInfoKHR<'_> {}
@@ -12940,13 +12940,13 @@ unsafe impl<'a> TaggedStructure for D3D1
 unsafe impl ExtendsSubmitInfo for D3D12FenceSubmitInfoKHR<'_> {}
 impl<'a> D3D12FenceSubmitInfoKHR<'a> {
     #[inline]
-    pub fn wait_semaphore_values(mut self, wait_semaphore_values: &'a [u64]) -> Self {
+    pub fn wait_semaphore_values(mut self, wait_semaphore_values: &'a [usize]) -> Self {
         self.wait_semaphore_values_count = wait_semaphore_values.len() as _;
         self.p_wait_semaphore_values = wait_semaphore_values.as_ptr();
         self
     }
     #[inline]
-    pub fn signal_semaphore_values(mut self, signal_semaphore_values: &'a [u64]) -> Self {
+    pub fn signal_semaphore_values(mut self, signal_semaphore_values: &'a [usize]) -> Self {
         self.signal_semaphore_values_count = signal_semaphore_values.len() as _;
         self.p_signal_semaphore_values = signal_semaphore_values.as_ptr();
         self
@@ -21911,9 +21911,9 @@ pub struct TimelineSemaphoreSubmitInfo<'
     pub s_type: StructureType,
     pub p_next: *const c_void,
     pub wait_semaphore_value_count: u32,
-    pub p_wait_semaphore_values: *const u64,
+    pub p_wait_semaphore_values: *const usize,
     pub signal_semaphore_value_count: u32,
-    pub p_signal_semaphore_values: *const u64,
+    pub p_signal_semaphore_values: *const usize,
     pub _marker: PhantomData<&'a ()>,
 }
 unsafe impl Send for TimelineSemaphoreSubmitInfo<'_> {}
@@ -21939,13 +21939,13 @@ unsafe impl ExtendsSubmitInfo for Timeli
 unsafe impl ExtendsBindSparseInfo for TimelineSemaphoreSubmitInfo<'_> {}
 impl<'a> TimelineSemaphoreSubmitInfo<'a> {
     #[inline]
-    pub fn wait_semaphore_values(mut self, wait_semaphore_values: &'a [u64]) -> Self {
+    pub fn wait_semaphore_values(mut self, wait_semaphore_values: &'a [usize]) -> Self {
         self.wait_semaphore_value_count = wait_semaphore_values.len() as _;
         self.p_wait_semaphore_values = wait_semaphore_values.as_ptr();
         self
     }
     #[inline]
-    pub fn signal_semaphore_values(mut self, signal_semaphore_values: &'a [u64]) -> Self {
+    pub fn signal_semaphore_values(mut self, signal_semaphore_values: &'a [usize]) -> Self {
         self.signal_semaphore_value_count = signal_semaphore_values.len() as _;
         self.p_signal_semaphore_values = signal_semaphore_values.as_ptr();
         self
@@ -21962,7 +21962,7 @@ pub struct SemaphoreWaitInfo<'a> {
     pub flags: SemaphoreWaitFlags,
     pub semaphore_count: u32,
     pub p_semaphores: *const Semaphore,
-    pub p_values: *const u64,
+    pub p_values: *const usize,
     pub _marker: PhantomData<&'a ()>,
 }
 unsafe impl Send for SemaphoreWaitInfo<'_> {}
@@ -21997,7 +21997,7 @@ impl<'a> SemaphoreWaitInfo<'a> {
         self
     }
     #[inline]
-    pub fn values(mut self, values: &'a [u64]) -> Self {
+    pub fn values(mut self, values: &'a [usize]) -> Self {
         self.semaphore_count = values.len() as _;
         self.p_values = values.as_ptr();
         self
@@ -22012,7 +22012,7 @@ pub struct SemaphoreSignalInfo<'a> {
     pub s_type: StructureType,
     pub p_next: *const c_void,
     pub semaphore: Semaphore,
-    pub value: u64,
+    pub value: usize,
     pub _marker: PhantomData<&'a ()>,
 }
 unsafe impl Send for SemaphoreSignalInfo<'_> {}
@@ -22024,7 +22024,7 @@ impl ::core::default::Default for Semaph
             s_type: Self::STRUCTURE_TYPE,
             p_next: ::core::ptr::null(),
             semaphore: Semaphore::default(),
-            value: u64::default(),
+            value: usize::default(),
             _marker: PhantomData,
         }
     }
@@ -22039,7 +22039,7 @@ impl<'a> SemaphoreSignalInfo<'a> {
         self
     }
     #[inline]
-    pub fn value(mut self, value: u64) -> Self {
+    pub fn value(mut self, value: usize) -> Self {
         self.value = value;
         self
     }
--- firefox-158.0/third_party/rust/ash/src/vk/extensions.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/ash/src/vk/extensions.rs	2026-10-13 19:47:59.069112412 +0200
@@ -5945,7 +5945,7 @@ pub const KHR_TIMELINE_SEMAPHORE_SPEC_VE
 pub type PFN_vkGetSemaphoreCounterValue = unsafe extern "system" fn(
     device: crate::vk::Device,
     semaphore: Semaphore,
-    p_value: *mut u64,
+    p_value: *mut usize,
 ) -> Result;
 #[allow(non_camel_case_types)]
 pub type PFN_vkWaitSemaphores = unsafe extern "system" fn(
--- firefox-158.0/third_party/rust/wgpu-core/src/command/ray_tracing.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-core/src/command/ray_tracing.rs	2026-10-13 19:48:28.094760391 +0200
@@ -2,7 +2,7 @@ use alloc::{sync::Arc, vec::Vec};
 use core::{
     cmp::max,
     mem,
-    num::NonZeroU64,
+    num::NonZeroUsize,
     ops::{Deref, Range},
 };
 
@@ -490,7 +490,7 @@ impl CommandBufferMutable {
         for action in &self.as_actions {
             match action {
                 AsAction::Build(build) => {
-                    let build_command_index = NonZeroU64::new(
+                    let build_command_index = NonZeroUsize::new(
                         command_index_guard.next_acceleration_structure_build_command_index,
                     )
                     .unwrap();
--- firefox-158.0/third_party/rust/wgpu-core/src/device/queue.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-core/src/device/queue.rs	2026-10-13 19:47:59.075971851 +0200
@@ -4,7 +4,7 @@ use alloc::{boxed::Box, string::String,
 use core::{
     iter,
     mem::{self, ManuallyDrop},
-    num::NonZeroU64,
+    num::NonZeroUsize,
     ptr::NonNull,
     sync::atomic::Ordering,
 };
@@ -240,7 +240,7 @@ impl Queue {
     /// [`Device::command_indices`]
     pub(crate) fn maintain(
         &self,
-        submission_index: u64,
+        submission_index: usize,
         snatch_guard: &SnatchGuard,
     ) -> (
         SmallVec<[SubmittedWorkDoneClosure; 1]>,
@@ -2081,7 +2081,7 @@ impl Queue {
 
         command_indices_lock.next_acceleration_structure_build_command_index += 1;
         let built_index =
-            NonZeroU64::new(command_indices_lock.next_acceleration_structure_build_command_index)
+            NonZeroUsize::new(command_indices_lock.next_acceleration_structure_build_command_index)
                 .unwrap();
 
         let new_blas = Arc::new(Blas {
--- firefox-158.0/third_party/rust/wgpu-core/src/device/resource.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-core/src/device/resource.rs	2026-10-13 19:48:28.096922562 +0200
@@ -67,7 +67,7 @@ use super::{
     ENTRYPOINT_FAILURE_ERROR, ZERO_BUFFER_SIZE,
 };
 
-use wgpu_sync::atomic::AtomicU64;
+use wgpu_sync::atomic::AtomicUsize;
 
 pub(crate) struct CommandIndices {
     /// The index of the last command submission that was attempted.
@@ -78,7 +78,7 @@ pub(crate) struct CommandIndices {
     ///
     /// [`last_successful_submission_index`]: Device::last_successful_submission_index
     pub(crate) active_submission_index: hal::FenceValue,
-    pub(crate) next_acceleration_structure_build_command_index: u64,
+    pub(crate) next_acceleration_structure_build_command_index: usize,
 }
 
 /// Parameters provided to shaders via a uniform buffer of the type
@@ -613,7 +613,7 @@ impl Device {
                         next_acceleration_structure_build_command_index: 1,
                     },
                 ),
-                last_successful_submission_index: AtomicU64::new(0),
+                last_successful_submission_index: AtomicUsize::new(0),
                 fence: ManuallyDrop::new(fence),
                 snatchable_lock: unsafe { SnatchLock::new(rank::DEVICE_SNATCHABLE_LOCK) },
                 valid: AtomicBool::new(true),
--- firefox-158.0/third_party/rust/wgpu-core/src/resource.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-core/src/resource.rs	2026-10-13 19:47:59.070619842 +0200
@@ -3,7 +3,7 @@ use core::{
     borrow::Borrow,
     fmt,
     mem::{self, size_of, ManuallyDrop},
-    num::NonZeroU64,
+    num::NonZeroUsize,
     ops::Range,
     ptr::NonNull,
 };
@@ -3288,7 +3288,7 @@ pub struct Blas {
     pub(crate) sizes: wgt::BlasGeometrySizeDescriptors,
     pub(crate) flags: wgt::AccelerationStructureFlags,
     pub(crate) update_mode: wgt::AccelerationStructureUpdateMode,
-    pub(crate) built_index: RwLock<Option<NonZeroU64>>,
+    pub(crate) built_index: RwLock<Option<NonZeroUsize>>,
     pub(crate) handle: u64,
     /// The `label` from the descriptor used to create the resource.
     pub(crate) label: String,
@@ -3524,7 +3524,7 @@ pub struct Tlas {
     pub(crate) max_instance_count: u32,
     pub(crate) flags: wgt::AccelerationStructureFlags,
     pub(crate) update_mode: wgt::AccelerationStructureUpdateMode,
-    pub(crate) built_index: RwLock<Option<NonZeroU64>>,
+    pub(crate) built_index: RwLock<Option<NonZeroUsize>>,
     pub(crate) dependencies: RwLock<Vec<Arc<Blas>>>,
     /// The `label` from the descriptor used to create the resource.
     pub(crate) label: String,
--- firefox-158.0/third_party/rust/wgpu-hal/src/lib.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-hal/src/lib.rs	2026-10-13 19:48:28.099625957 +0200
@@ -326,8 +326,8 @@ pub const QUERY_SIZE: wgt::BufferAddress
 
 pub type Label<'a> = Option<&'a str>;
 pub type MemoryRange = Range<wgt::BufferAddress>;
-pub type FenceValue = u64;
-pub type AtomicFenceValue = wgpu_sync::atomic::AtomicU64;
+pub type FenceValue = usize;
+pub type AtomicFenceValue = wgpu_sync::atomic::AtomicUsize;
 
 /// A callback to signal that wgpu is no longer using a resource.
 #[cfg(all(any(gles, vulkan, metal), not(webgl)))]
--- firefox-158.0/third_party/rust/wgpu-hal/src/metal/mod.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-hal/src/metal/mod.rs	2026-10-13 19:47:59.071178508 +0200
@@ -532,7 +532,7 @@ impl Queue {
             .ok_or(crate::DeviceError::OutOfMemory)?;
         let _ = self.shared.relay.set(Relay {
             event,
-            next_release_value: atomic::AtomicU64::new(1),
+            next_release_value: atomic::AtomicUsize::new(1),
             commit_lock: Mutex::new(()),
         });
         Ok(())
@@ -612,7 +612,7 @@ type PendingEvents = Mutex<Vec<(Retained
 #[derive(Debug)]
 struct Relay {
     event: Retained<ProtocolObject<dyn MTLSharedEvent>>,
-    next_release_value: atomic::AtomicU64,
+    next_release_value: atomic::AtomicUsize,
     commit_lock: Mutex<()>,
 }
 
--- firefox-158.0/third_party/rust/wgpu-hal/src/noop/mod.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-hal/src/noop/mod.rs	2026-10-13 19:48:28.097676486 +0200
@@ -3,7 +3,7 @@
 use alloc::{string::String, sync::Arc, vec, vec::Vec};
 use core::{ptr, sync::atomic::Ordering, time::Duration};
 
-use wgpu_sync::atomic::AtomicU64;
+use wgpu_sync::atomic::AtomicUsize;
 
 use crate::TlasInstance;
 
@@ -26,7 +26,7 @@ pub struct Resource;
 
 #[derive(Debug)]
 pub struct Fence {
-    value: AtomicU64,
+    value: AtomicUsize,
 }
 
 type DeviceResult<T> = Result<T, crate::DeviceError>;
@@ -424,7 +424,7 @@ impl crate::Device for Context {
     unsafe fn destroy_query_set(&self, set: Resource) {}
     unsafe fn create_fence(&self) -> DeviceResult<Fence> {
         Ok(Fence {
-            value: AtomicU64::new(0),
+            value: AtomicUsize::new(0),
         })
     }
     unsafe fn destroy_fence(&self, fence: Fence) {}
--- firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/mod.rs	2026-10-08 04:24:20.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/mod.rs	2026-10-13 19:47:59.076891635 +0200
@@ -1662,7 +1662,7 @@ impl Queue {
         &self.device.raw
     }
 
-    pub fn add_signal_semaphore(&self, semaphore: vk::Semaphore, semaphore_value: Option<u64>) {
+    pub fn add_signal_semaphore(&self, semaphore: vk::Semaphore, semaphore_value: Option<usize>) {
         let mut guard = self.signal_semaphores.lock();
         if let Some(value) = semaphore_value {
             guard.push_signal(SemaphoreType::Timeline(semaphore, value));
@@ -1693,7 +1693,7 @@ impl Queue {
     pub fn add_wait_semaphore(
         &self,
         semaphore: vk::Semaphore,
-        semaphore_value: Option<u64>,
+        semaphore_value: Option<usize>,
         stage: vk::PipelineStageFlags,
     ) {
         let mut guard = self.wait_semaphores.lock();
--- firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/semaphore_list.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/semaphore_list.rs	2026-10-13 19:47:59.071910573 +0200
@@ -42,7 +42,7 @@ pub struct SemaphoreList {
     /// Vulkan ignores these.
     ///
     /// [`semaphores`]: Self::semaphores
-    values: Vec<u64>,
+    values: Vec<usize>,
 
     /// Stage masks for wait semaphores.
     ///
@@ -214,5 +214,5 @@ impl SemaphoreList {
 
 pub enum SemaphoreType {
     Binary(vk::Semaphore),
-    Timeline(vk::Semaphore, u64),
+    Timeline(vk::Semaphore, usize),
 }
--- firefox-158.0/third_party/rust/wgpu-types/src/lib.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-types/src/lib.rs	2026-10-13 19:47:59.076624333 +0200
@@ -250,7 +250,7 @@ pub enum PollError {
     /// The requested Wait timed out before the submission was completed.
     Timeout,
     /// The requested Wait was given a wrong submission index.
-    WrongSubmissionIndex(u64, u64),
+    WrongSubmissionIndex(usize, usize),
 }
 
 // This impl could be derived by `thiserror`, but by not doing so, we can reduce the number of
--- firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/swapchain/native.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/swapchain/native.rs	2026-10-13 19:47:59.076716246 +0200
@@ -958,7 +958,7 @@ struct NativeSwapchainSubmissionSemaphor
 }
 
 impl<'a> SwapchainSubmissionSemaphoreGuard for NativeSwapchainSubmissionSemaphoreGuard<'a> {
-    fn set_used_fence_value(&mut self, value: u64) {
+    fn set_used_fence_value(&mut self, value: usize) {
         self.acquire_semaphore_guard.set_used_fence_value(value);
     }
 
--- firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/swapchain/mod.rs	2026-10-08 04:24:21.000000000 +0200
+++ firefox-158.0/third_party/rust/wgpu-hal/src/vulkan/swapchain/mod.rs	2026-10-13 19:47:59.076776470 +0200
@@ -106,7 +106,7 @@ pub(super) trait SurfaceTextureMetadata:
 /// Guard type for managing swapchain submission semaphores.
 pub(super) trait SwapchainSubmissionSemaphoreGuard {
     /// Sets the Fence value for this submission.
-    fn set_used_fence_value(&mut self, value: u64);
+    fn set_used_fence_value(&mut self, value: usize);
 
     /// Gets semaphores to wait on before doing GPU work for this swapchain texture.
     fn get_acquire_wait_semaphore(&mut self) -> Option<SemaphoreType>;
--- firefox-158.0/third_party/rust/wgpu-core-remote-types/src/lib.rs
+++ firefox-158.0/third_party/rust/wgpu-core-remote-types/src/lib.rs
@@ -19,7 +19,7 @@ use alloc::borrow::Cow;
 
 pub type Index = u32;
 pub type Epoch = u32;
-pub type SubmissionIndex = u64;
+pub type SubmissionIndex = usize;
 pub type SubmittedWorkDoneClosure = Box<dyn FnOnce() + Send + 'static>;
 
 pub mod id;

--- firefox-158.0/gfx/wgpu_bindings/src/server.rs
+++ firefox-158.0/gfx/wgpu_bindings/src/server.rs
@@ -3201,7 +3201,7 @@ pub unsafe extern "C" fn wgpu_server_que
     let _ = signal_semaphores;
 
     submission_succeeded
-        .then_some(submission_index)
+        .then_some(submission_index as u64)
         .unwrap_or(0)
 }
